app/master/build.py

```diff
@@ -564,6 +569,121 @@ def _generate_unique_symlink_path_for_build_repo(self):
         """
         return os.path.join(Configuration['build_symlink_directory'], str(uuid.uuid4()))
 
+    def save(self):
+        """Serialize the Build object and update all of the parts to the database."""
+        session = Connection.get()
+
+        q_build = session.query(BuildSchema).filter(BuildSchema.build_id == self._build_id).first()
+        q_failed_artifact_directories = session.query(FailedArtifactDirectoriesSchema).filter(FailedArtifactDirectoriesSchema.build_id == self._build_id).all()
+        q_failed_subjob_atom_pairs = session.query(FailedSubjobAtomPairsSchema).filter(FailedSubjobAtomPairsSchema.build_id == self._build_id).all()
+        q_build_atoms = session.query(AtomsSchema).filter(AtomsSchema.build_id == self._build_id).all()
+        q_build_subjobs = session.query(SubjobsSchema).filter(SubjobsSchema.build_id == self._build_id).all()
+
+        # If this wasn't found, it's safe to assume that the build doesn't exist within the database
+        if q_build is None:
+            raise ItemNotFoundError('Unable to find build (id: {}) in database.'.format(self._build_id))
+
+        q_build.completed = self._status() == BuildState.FINISHED
+
+        q_build.artifacts_tar_file = self._artifacts_tar_file
+        q_build.artifacts_zip_file = self._artifacts_zip_file
+        q_build.error_message = self._error_message
+        q_build.postbuild_tasks_are_finished = self._postbuild_tasks_are_finished
+        q_build.setup_failures = self.setup_failures
+        q_build.timing_file_path = self._timing_file_path
+
+        build_artifact_dir = None
+        if self._build_artifact is not None:
+            build_artifact_dir = self._build_artifact.build_artifact_dir
+
+        q_build.build_artifact_dir = build_artifact_dir
+
+        if self._build_artifact is not None:
+            # Clear all old directories
+            session.query(FailedArtifactDirectoriesSchema) \
+                .filter(FailedArtifactDirectoriesSchema.build_id == self._build_id) \
+                .delete()
+
+            # Commit changes so we don't delete the newly added rows later
+            session.commit()
+
+            # Add all the updated versions of the directories
+            for directory in self._build_artifact._get_failed_artifact_directories():  # pylint: disable=protected-access
```

Comment:
Private access :( -- how do you feel about letting the BuildArtifact class (and other table-related classes) store/load its own state?

On a higher level, how do you feel about not storing any internals of the BuildArtifact class at all? Its constructor only takes in a directory on disk, and it seems like all of its internal state can be computed based on that. (Maybe it would take some refactoring to get that working, but that's not surprising for this change. I'd rather do minor refactoring of the code than add a bunch of extra columns to the db.)